README.md
# [pw_edit_field](https://github.com/lisen87/pw_edit_field)

flutter 密码输入框

flutter password input box

> Supported  Platforms
> * Android
> * iOS

## How to Use

```yaml
# add this line to your dependencies
pw_edit_field: ^2.0.0+1
```

```dart
import 'package:pw_edit_field/pw_edit_field.dart';
```

```dart
///输入完成后自动关闭输入法，并返回输入值
///Automatically close the input method after the input is completed, and return the input value

 PWEditField(
    textFieldCount: 6,
    onInputDone: (pw){
      
    },
  ),
  
  ///输入后，通过controller.text获取输入值
  ///After input, get the input value through controller.text
  
  PWEditField(
     textFieldCount: 3,
     controller: _controller,
   ),
  
```
![](https://github.com/lisen87/pw_edit_field/blob/master/screenshots/screenshot_001.png)


CHANGELOG.md
## [2.0.0+1]
* 输入法弹出问题
## [2.0.0]
* 空安全
## [1.0.1+2]
* 处理输入框遮挡问题
* Handle input box occlusion
## [1.0.1+1]
* 保持输入状态
* Keep input
## [1.0.1]
* 增加输入监听
* Increase input monitoring

## [1.0.0]
* flutter Password input box, pure dart language


lib/pw_edit_field.dart
import 'package:flutter/cupertino.dart';
import 'package:flutter/material.dart';
import 'package:flutter/services.dart';

/// Created by liSen on 2019/12/7 15:00.
/// @author liSen < 453354858@qq.com >

class PWEditField extends StatefulWidget {
  ///所有密码框输入完成后监听 Listen after all password boxes are entered
  ///
  final ValueChanged<String>? onInputDone;

  final ValueChanged<String>? onChanged;

  ///密码框数量 Number of password boxes
  ///
  final int? textFieldCount;

  ///密码框宽度 Password box width
  ///
  final double textFieldWidth;

  ///密码框高度 Password box height
  ///
  final double textFieldHeight;

  ///密码框之间的间距 Space between password boxes
  ///
  final double textFieldSpace;

  ///以下是TextField 原有属性 Following are the original TextField properties
  ///
  final bool autoFocus;
  final bool obscureText;
  final bool isDoneClean;
  final TextStyle textStyle;
  final TextInputType keyboardType;
  final List<TextInputFormatter>? inputFormatters;
  BoxDecoration? focusDecoration;
  BoxDecoration? unFocusDecoration;
  TextEditingController? controller;

  /// 只能输入数字 Only numbers can be entered
  /// inputFormatters = [WhitelistingTextInputFormatter(RegExp("[0-9]")),]

  PWEditField({
    @required this.textFieldCount,
    this.textFieldWidth = 30,
    this.textFieldHeight = 30,
    this.textFieldSpace = 3,
    this.autoFocus = false,
    this.obscureText = true,
    this.isDoneClean = false,
    this.textStyle = const TextStyle(
      fontSize: 12,
      color: Colors.black54,
      decoration: TextDecoration.none,
    ),
    this.keyboardType = TextInputType.number,
    this.inputFormatters,
    this.focusDecoration,
    this.unFocusDecoration,
    this.onInputDone,
    this.onChanged,
    this.controller,
    Key? key,
  }) : super(
          key: key,
        ) {
    this.controller ??= TextEditingController();
    this.focusDecoration ??=
        BoxDecoration(
          border: Border.all(color: Colors.blue, width: 1),
          borderRadius: BorderRadius.all(Radius.circular(2)),
        );
    this.unFocusDecoration ??=
        BoxDecoration(
          border: Border.all(color: Colors.grey, width: 1),
          borderRadius: BorderRadius.all(Radius.circular(2)),
        );
  }

  @override
  State<StatefulWidget> createState() {
    return PWEditFieldState();
  }
}

class PWEditFieldState extends State<PWEditField> {
  FocusNode _focusNode = FocusNode();
  int _currentIndex = -1;

  @override
  void initState() {
    super.initState();
    _currentIndex = widget.autoFocus ? 0 : -1;
    _focusNode.addListener(() {
      if (!_focusNode.hasFocus) {
        _currentIndex = -1;
        setState(() {});
      }
    });
  }

  @override
  Widget build(BuildContext context) {
    double w = widget.textFieldCount! * widget.textFieldWidth +
        widget.textFieldCount! * widget.textFieldSpace * 2;
    return SingleChildScrollView(
      scrollDirection: Axis.horizontal,
      child: Stack(
        alignment: AlignmentDirectional.bottomStart,
        children: <Widget>[
          getTextField(),
          Container(
            width: w,
            height: widget.textFieldHeight,
            child: Row(
              mainAxisAlignment: MainAxisAlignment.center,
              children: getChildren(),
            ),
          ),
        ],
      ),
    );
  }

  List<Widget> getChildren() {
    List<Widget> list = [];
    for (int i = 0; i < widget.textFieldCount!; i++) {
      list.add(getChild(i));
    }
    return list;
  }

  Widget getChild(int index) {
    return Padding(
      padding: EdgeInsets.symmetric(horizontal: widget.textFieldSpace),
      child: GestureDetector(
        onTap: () {
          _focusNode.unfocus();
          Future.delayed(Duration(microseconds: 100),(){
            FocusScope.of(context).requestFocus(_focusNode);
            _currentIndex = widget.controller?.text.length ?? 0 ;
            setState(() {});
          });
        },
        child: Container(
          decoration: _currentIndex == index
              ? widget.focusDecoration
              : widget.unFocusDecoration,
          width: widget.textFieldWidth,
          height: widget.textFieldHeight,
          child: Center(
            child: getTextWidget(index),
          ),
        ),
      ),
    );
  }

  Widget getTextWidget(int index) {
    String data = "";
    String? text = widget.controller?.text;
    if (text != null && text.length > index) {
      data = widget.obscureText ? "●" : text.substring(index, index + 1);
    }
    return Text(
      data,
      maxLines: 1,
      style: widget.textStyle,
      textAlign: TextAlign.center,
    );
  }

  Widget getTextField() {
    return Container(
      width: 0,
      height: 0,
      child: TextField(
        enableInteractiveSelection: false,
        focusNode: _focusNode,
//        showCursor: false,
        autofocus: widget.autoFocus,
        controller: widget.controller,
        maxLength: widget.textFieldCount,
        maxLines: 1,
        obscureText: widget.obscureText,
        style: widget.textStyle,
        keyboardType: widget.keyboardType,
        inputFormatters: widget.inputFormatters,
        decoration: InputDecoration(
          border: InputBorder.none,
          counterText: "",
          contentPadding: EdgeInsets.only(
            top: 0,
            bottom: 0,
            left: 2,
          ),
        ),
        textAlign: TextAlign.center,
        onChanged: (String text) {
          if (widget.onChanged != null) {
            widget.onChanged!(text);
          }
          _currentIndex = text.length;
          setState(() {});
          if (widget.onInputDone != null &&
              text.length == widget.textFieldCount) {
            widget.onInputDone!(widget.controller!.text);
            _focusNode.unfocus();
            if(widget.isDoneClean){
              widget.controller!.text = "";
            }
          }
        },
      ),
    );
  }

  @override
  void dispose() {
    _focusNode.dispose();
    super.dispose();
  }
}


lib/PasswordWidget.dart
import 'package:flutter/cupertino.dart';
import 'package:flutter/material.dart';
import 'package:flutter/services.dart';

/// Created by liSen on 2019/12/7 15:00.
/// @author liSen < 453354858@qq.com >

class PasswordWidget extends StatefulWidget {
  ///所有密码框输入完成后监听 Listen after all password boxes are entered
  ///
  final ValueChanged<String> onInputDone;

  ///密码框数量 Number of password boxes
  ///
  final int textFieldCount;

  ///密码框宽度 Password box width
  ///
  final double textFieldWidth;

  ///密码框高度 Password box height
  ///
  final double textFieldHeight;

  ///密码框之间的间距 Space between password boxes
  ///
  final double textFieldSpace;

  ///以下是TextField 原有属性 Following are the original TextField properties
  ///
  final bool autoFocus;
  final bool obscureText;
  final TextStyle textStyle;
  final TextInputType keyboardType;
  final List<TextInputFormatter> inputFormatters;
  final BoxDecoration focusDecoration;
  final BoxDecoration unFocusDecoration;

  /// 只能输入数字 Only numbers can be entered
  /// inputFormatters = [WhitelistingTextInputFormatter(RegExp("[0-9]")),]

  PasswordWidget({
    @required this.textFieldCount,
    this.textFieldWidth = 30,
    this.textFieldHeight = 30,
    this.textFieldSpace = 3,
    this.autoFocus = false,
    this.obscureText = true,
    this.textStyle = const TextStyle(
      fontSize: 12,
      color: Colors.black54,
    ),
    this.keyboardType = TextInputType.number,
    this.inputFormatters,
    Decoration focusDecoration,
    Decoration unFocusDecoration,
    this.onInputDone,
  })  : focusDecoration = focusDecoration ??
            BoxDecoration(
              border: Border.all(color: Colors.blue, width: 1),
              borderRadius: BorderRadius.all(Radius.circular(2)),
            ),
        unFocusDecoration = unFocusDecoration ??
            BoxDecoration(
              border: Border.all(color: Colors.grey, width: 1),
              borderRadius: BorderRadius.all(Radius.circular(2)),
            );

  @override
  State<StatefulWidget> createState() {
    return PasswordWidgetState();
  }
}

class PasswordWidgetState extends State<PasswordWidget> {
  TextEditingController _controller = TextEditingController();
  FocusNode _focusNode = FocusNode();
  int _currentIndex = -1;
  @override
  void initState() {
    super.initState();
    _currentIndex = widget.autoFocus ? 0 : -1;
    _focusNode.addListener(() {
      if (!_focusNode.hasFocus) {
        _currentIndex = -1;
        setState(() {});
      }
    });
  }

  @override
  Widget build(BuildContext context) {
    double w = widget.textFieldCount * widget.textFieldWidth +
        widget.textFieldCount * widget.textFieldSpace * 2;
    return SingleChildScrollView(
      scrollDirection: Axis.horizontal,
      child: Stack(
        children: <Widget>[
          getTextField(),
          Container(
            width: w,
            height: widget.textFieldHeight,
            child: Row(
              mainAxisAlignment: MainAxisAlignment.center,
              children: getChildren(),
            ),
          ),
        ],
      ),
    );
  }

  List<Widget> getChildren() {
    List<Widget> list = [];
    for (int i = 0; i < widget.textFieldCount; i++) {
      list.add(getChild(i));
    }
    return list;
  }

  Widget getChild(int index) {
    return Padding(
      padding: EdgeInsets.symmetric(horizontal: widget.textFieldSpace),
      child: GestureDetector(
        onTap: () {
          _focusNode.unfocus();
          FocusScope.of(context).requestFocus(_focusNode);
          _currentIndex = _controller.text.length;
          setState(() {});
        },
        child: Container(
          decoration: _currentIndex == index
              ? widget.focusDecoration
              : widget.unFocusDecoration,
          width: widget.textFieldWidth,
          height: widget.textFieldHeight,
          child: Center(
            child: getTextWidget(index),
          ),
        ),
      ),
    );
  }

  Widget getTextWidget(int index) {
    String data = "";
    String text = _controller.text;
    if (text.length > index) {
      data = widget.obscureText ? "●" : text.substring(index, index + 1);
    }
    return Text(
      data,
      maxLines: 1,
      style: widget.textStyle,
      textAlign: TextAlign.center,
    );
  }

  Widget getTextField() {
    return Container(
      width: 0,
      height: 0,
      child: TextField(
        enableInteractiveSelection: false,
        focusNode: _focusNode,
        showCursor: false,
        autofocus: widget.autoFocus,
        controller: _controller,
        maxLength: widget.textFieldCount,
        maxLines: 1,
        obscureText: widget.obscureText,
        style: widget.textStyle,
        keyboardType: widget.keyboardType,
        inputFormatters: widget.inputFormatters,
        decoration: InputDecoration(
          border: InputBorder.none,
          counterText: "",
          contentPadding: EdgeInsets.only(
            top: 0,
            bottom: 0,
            left: 2,
          ),
        ),
        textAlign: TextAlign.center,
        onChanged: (String text) {
          _currentIndex = text.length;
          setState(() {});
          if (widget.onInputDone != null && text.length == widget.textFieldCount) {
            widget.onInputDone(_controller.text);
            _controller.text = "";
            _focusNode.unfocus();
          }
        },
      ),
    );
  }

  @override
  void dispose() {
    _controller.dispose();
    _focusNode.dispose();
    super.dispose();
  }
}


example/lib/main.dart
import 'package:flutter/cupertino.dart';
import 'package:flutter/material.dart';
import 'package:flutter/services.dart';
import 'package:pw_edit_field/pw_edit_field.dart';
import 'dart:ui' as ui;
import 'package:flutter/foundation.dart';

void main() => runApp(MyApp());

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(
        primarySwatch: Colors.blue,
      ),
      home: MyHomePage(title: 'Flutter Demo Home Page'),
    );
  }
}

class MyHomePage extends StatefulWidget {
  MyHomePage({Key? key, this.title}) : super(key: key);

  final String? title;

  @override
  _MyHomePageState createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  String pw = "";

  TextEditingController _controller = TextEditingController();

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text(widget.title!),
      ),
      body: SingleChildScrollView(
        child: Center(
          child: Column(
            mainAxisAlignment: MainAxisAlignment.center,
            children: <Widget>[
              SizedBox(
                height: 200,
              ),
              GestureDetector(
                  onTap: () {
                    showDialog(
                        context: context,
                        barrierDismissible: false,
                        builder: (BuildContext context) {
                          var mediaQueryData =
                              MediaQueryData.fromWindow(ui.window);
                          return CupertinoAlertDialog(
                            content: Container(
                              decoration: BoxDecoration(
                                  color: Colors.white,
                                  borderRadius:
                                      BorderRadius.all(Radius.circular(4))),
                              child: Center(
                                child: Column(
                                  mainAxisSize: MainAxisSize.min,
                                  children: <Widget>[
                                    Container(
                                      height: 200,
                                      color: Colors.teal,
                                    ),
                                    Container(
                                      child: Material(
                                          child: AnimatedContainer(
                                              duration: const Duration(
                                                  milliseconds: 300),
                                              padding: EdgeInsets.only(
                                                  bottom: mediaQueryData
                                                      .viewInsets.bottom),
                                              child: PWEditField(
                                                textFieldCount: 5,
                                                controller: _controller,
                                              ))),
                                    ),
                                  ],
                                ),
                              ),
                            ),
                          );
                        });
                  },
                  child: Text("密码 ：" + pw)),
              SizedBox(
                height: 8,
              ),
              PWEditField(
                textFieldCount: 8,
                textFieldSpace: 5,
                textFieldWidth: 30,
                textFieldHeight: 30,
                obscureText: false,
                textStyle: TextStyle(fontSize: 18),
                keyboardType: TextInputType.number,
                inputFormatters: [
                  FilteringTextInputFormatter(RegExp("[0-9]"), allow: true),
                ],
                focusDecoration: BoxDecoration(
                  border:
                      Border(bottom: BorderSide(color: Colors.teal, width: 2)),
                ),
                unFocusDecoration: BoxDecoration(
                  border:
                      Border(bottom: BorderSide(color: Colors.red, width: 2)),
                ),
                onInputDone: (pw) {
                  this.pw = pw;

                  setState(() {});
                },
              ),
              SizedBox(
                height: 8,
              ),
              PWEditField(
                textFieldCount: 6,
                textFieldWidth: 40,
                textFieldHeight: 40,
                inputFormatters: [
                  FilteringTextInputFormatter(RegExp("[0-9]"), allow: true),
                ],
                autoFocus: true,
                onInputDone: (pw) {
                  this.pw = pw;
                  setState(() {
//                  _text.data = pw;
                  });
                },
              ),
              SizedBox(
                height: 8,
              ),
              PWEditField(
                textFieldCount: 6,
                onInputDone: (pw) {
                  this.pw = pw;
                  setState(() {});
                },
              ),
              PWEditField(
                textFieldCount: 3,
                controller: _controller,
                onInputDone: (text) {
                  print(text);
                },
              ),
              Padding(
                padding: const EdgeInsets.all(8.0),
                child: PWEditField(
                  textFieldCount: 8,
                  textFieldSpace: 10,
                  textFieldWidth: 40,
                  textFieldHeight: 20,
                  obscureText: false,
                  textStyle: TextStyle(fontSize: 15),
                  keyboardType: TextInputType.visiblePassword,
                  focusDecoration: BoxDecoration(
                    border: Border.all(color: Colors.teal, width: 1),
                    borderRadius: BorderRadius.all(Radius.circular(5)),
                  ),
                  unFocusDecoration: BoxDecoration(
                    border: Border.all(color: Colors.blueGrey, width: 1),
                    borderRadius: BorderRadius.all(Radius.circular(5)),
                  ),
                  onInputDone: (pw) {
                    this.pw = pw;
                    setState(() {});
                  },
                ),
              ),
            ],
          ),
        ),
      ),
    );
  }
}
